### PR TITLE
Univ3 router

### PR DESCRIPTION
**Pull request by Max-RiskHarbor**
Pretty simple implementation using virtual reserves interpretation of univ3 so we can reuse the constant product find arb code when possible.

### CODE REVIEW

**Comment by Max-RiskHarbor** [comment]
May want to make an abstract constant product pool to accommodate other versions of concentrated liquidity.

File: src/cfmms.jl [hunk 1]
```diff
@@ -190,3 +190,113 @@ function find_arb!(Δ::VT, Λ::VT, cfmm::GeometricMeanTwoCoin{T}, v::VT) where {
...
+        )
+    end
+end
+
+## See univ3 whitepaper
+function virtual_reserves(P,L)
+    sP = sqrt(P)
+    x = L/sP
+    y = L*sP
+    return([x, y])
```
Comment (tjdiamandis) [inline src/cfmms.jl:218]: change to return x, y (avoid creating an array)

File: examples/Univ3.jl [hunk 2]
```diff
@@ -0,0 +1,13 @@
+using CFMMRouter
+using LinearAlgebra, SparseArrays, StaticArrays
+include("../src/cfmms.jl")
+
+tick_list = [ Dict("price" => .1, "liquidity" => 1.0), Dict("price" => .5, "liquidity" => 1.0), Dict("price" => 1.0, "liquidity" => 100.0), Dict("price" => 2.0, "liquidity" => 1.0), Dict("price" => 4.0, "liquidity" => 1.0) ]
```
Comment (tjdiamandis) [inline examples/Univ3.jl:5]: Split into multiple lines + add a comment explaining how to construct

File: examples/Univ3.jl [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+
+pool = UniV3([100,100],1,[1,2],1.0,3,tick_list)
+Δ = [0.0,0.0]
+Λ = [0.0,0.0]
+
+find_arb!(Δ,Λ,pool, [1.0,2.0])
+
```
Comment (tjdiamandis) [inline examples/Univ3.jl:12]: print out the trade (& maybe the state of the pool as well)

File: src/cfmms.jl [hunk 4]
```diff
@@ -190,3 +190,113 @@ function find_arb!(Δ::VT, Λ::VT, cfmm::GeometricMeanTwoCoin{T}, v::VT) where {
     Λ[2] = geom_arb_λ(v[2]/v[1], R[2], R[1], η, γ)
     return nothing
 end
+
+
+struct UniV3{T} <: CFMM{T}
+    @add_two_coin_fields
+    current_price :: T
+    current_tick_index :: Int #This is the index of the maximal tick with price lower than current_price in the ticks dictionary
+    ticks #This is the tick mapping sorted by price
```
Comment (tjdiamandis) [inline src/cfmms.jl:199]: explain how mapping is constructed

File: src/cfmms.jl [hunk 5]
```diff
@@ -190,3 +190,113 @@ function find_arb!(Δ::VT, Λ::VT, cfmm::GeometricMeanTwoCoin{T}, v::VT) where {
...
+                Λ[2] += prod_arb_λ(1/prev_tick_price, R[2], k, 1)
+            end
+            i += 1
+        end
+    end
+    Δ = Δ ./ γ
+    Λ = Λ ./ γ
+    return nothing
+end
+
```
Comment (tjdiamandis) [inline src/cfmms.jl:302]: implement update reserves for UniV3 pool

File: src/cfmms.jl [hunk 6]
```diff
@@ -190,3 +190,113 @@ function find_arb!(Δ::VT, Λ::VT, cfmm::GeometricMeanTwoCoin{T}, v::VT) where {
...
+            i += 1
+        end
+
+    elseif target_price < current_price #iterate backwards in the tick mapping
+        i = 1
+        while true
+            prev_tick_price = ticks[current_tick_index - i + 1]["price"]
+            if prev_tick_price <= target_price ## so now we know that prev_tick_price < target_price <=  current_price
+
+                print(ticks[current_tick_index - i + 2]["price"])
```
Comment (tjdiamandis) [inline src/cfmms.jl:272]: remove print statements (or add `verbose` kwarg so all prints are changed to `verbose && print(...)`)

File: src/cfmms.jl [hunk 7]
```diff
@@ -190,3 +190,113 @@ function find_arb!(Δ::VT, Λ::VT, cfmm::GeometricMeanTwoCoin{T}, v::VT) where {
...
+
+struct UniV3{T} <: CFMM{T}
+    @add_two_coin_fields
+    current_price :: T
+    current_tick_index :: Int #This is the index of the maximal tick with price lower than current_price in the ticks dictionary
+    ticks #This is the tick mapping sorted by price
+    function UniV3(R,γ,idx,current_price,current_tick_index,ticks)
+        γ_T, idx_uint, T = two_coin_check_cast(R, γ, idx)
+        return new{T}(
+            MVector{2,T}(R),
```
Comment (tjdiamandis) [inline src/cfmms.jl:203]: Maybe reserves should be computed from the `tick_list`?

File: test/cfmms.jl [hunk 8]
```diff
@@ -70,6 +71,148 @@ end
...
+
+    # #reserves dont matter for univ3 pools, only tick data so just setting to 1
+    R = [1.0,1.0]
+    # #no fees for now
+    γ = .997
+    # # 
+    ids = [1.0,2.0]
+    current_price = 15.0
+    current_tick_index = 3.0
+    ticks = [Dict("price" => 1/2.0^64, "liquidity" => 0.0),Dict("price" => 5.0, "liquidity" => 1.0), Dict("price" => 10.0, "liquidity" => 2.0), Dict("price" => 20.0, "liquidity" => 1.0), Dict("price" => 30.0, "liquidity" => 0.0), Dict("price" => 2.0^64, "liquidity" => 0)]
```
Comment (tjdiamandis) [inline test/cfmms.jl:92]: line length

File: src/cfmms.jl [hunk 9]
```diff
@@ -190,3 +190,113 @@ function find_arb!(Δ::VT, Λ::VT, cfmm::GeometricMeanTwoCoin{T}, v::VT) where {
...
+        return nothing
+    end
+    target_price = target_price/γ
+    if target_price >= current_price #iterate forwards in the tick mapping
+        i = 1
+        while true
+            next_tick_price = ticks[current_tick_index + i]["price"]
+            if next_tick_price >= target_price ## so now we know that current_price <= target_price < next_tick_price
+                #print(max(current_price,ticks[current_tick_index + i - 1]["price"]))
+                R = virtual_reserves(max(current_price,ticks[current_tick_index + i - 1]["price"]),ticks[current_tick_index + i - 1]["liquidity"])
```
Comment (tjdiamandis) [inline src/cfmms.jl:241]: line length

**Review by tjdiamandis — CHANGES_REQUESTED** [review]

File: src/cfmms.jl (hunk 7 above)
Comment (Max-RiskHarbor) [reply src/cfmms.jl:203]: I think in practice the tick list alone doesn't tell you the reserves because there are fees accumulated at tick as well

Comment (tjdiamandis) [reply src/cfmms.jl:203]: Would we want to add those to the reserves in the tick list? I don't want duplicate info floating around where we have to remember to update both (or need to more formally define the CFMM inference to be implemented)

Comment (Max-RiskHarbor) [reply src/cfmms.jl:203]: ok so just checked on the status of how fees work and it turns out fees aren't immediately added to liquidity at tick, they need to be retrieved and then reinvested into the liquidity position before the changes to take effect. https://docs.uniswap.org/protocol/guides/providing-liquidity/collect-fees so I think the approach of just calculating the reserves from the tick mapping would work. But I'm not sure it is neccesary, I will implement it and then make it optional to construct this from the tick mapping if the reserves are not passed in.

Comment (tjdiamandis) [reply src/cfmms.jl:203]: Sounds good.

I think we should make the design choice to ignore fee accumulation. The onus is on the user to update the reserves appropriately before calculating a route.

**Review by tjdiamandis — COMMENTED** [review]
Thanks! Sorry for the delay on this

**Review by tjdiamandis — APPROVED** [review]

**Review by tjdiamandis — APPROVED** [review]

**Pull request merged**